### PR TITLE
Fix only_update_to_newer_versions regression

### LINKED ISSUE #6529: only_update_to_newer_versions regression in v1.16
Hi,

I've noticed a few edge cases where the `only_update_to_newer_versions` setting doesn't get respected anymore, where it used to work in Bundler 1.15.

Here is an example to reproduce: https://gist.github.com/theflow/edd84a5318738b59f65cf42eaaede9f3

Bundler v1.16 will downgrade carrierwave-mongoid to an really old version which is missing the version requirement on mongoid:

```
+ BUNDLE_ONLY_UPDATE_TO_NEWER_VERSIONS=true
+ bundle update mongoid carrierwave-mongoid
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies....
Using concurrent-ruby 1.0.5
Using i18n 1.0.1
Using minitest 5.11.3
Using thread_safe 0.3.6
Using tzinfo 1.2.5
Using activesupport 5.2.0
Using activemodel 5.2.0
Using bson 4.3.0
Using bundler 1.16.1
Using mime-types-data 3.2016.0521
Using mime-types 3.1
Using carrierwave 1.2.2
Using mongo 2.5.3
Using mongoid 7.0.1 (was 6.4.1)
Using carrierwave-mongoid 0.1.0 (was 1.0.0)
Bundler attempted to update carrierwave-mongoid but its version regressed from 1.0.0 to 0.1.0
Bundle updated!
```

I'm pretty sure it was introduced in https://github.com/bundler/bundler/commit/47256d20cb05ebc724ee67173094682153b6b4aa, but I don't understand the reason behind that change.

Here https://github.com/bundler/bundler/blob/master/lib/bundler/definition.rb#L975 it's comparing a `Bundler::Dependency` with an `Array` from the group_by which will always be non-equal, which I guess is not the intention?

Comparing the first element of the array fixes this issue for me, but not sure if that's the correct way and what's supposed to happen if there are multiple `dependencies_by_name`.

https://github.com/theflow/bundler/commit/07942b1a6ccec4faff8716afc89288edad8496d2

Cheers,
Florian

### PR DESCRIPTION
**Pull request by theflow**
This is my attempt to fix #6529

### What was the end-user problem that led to this PR?

Running `bundle update` with `BUNDLE_ONLY_UPDATE_TO_NEWER_VERSIONS: "true"` resulted in a gem getting downgraded to a really old version in a certain edge case. Ironically it wouldn't get downgraded when `BUNDLE_ONLY_UPDATE_TO_NEWER_VERSIONS` was set to false.

### What was your diagnosis of the problem?

My diagnosis was that https://github.com/bundler/bundler/commit/47256d20cb05ebc724ee67173094682153b6b4aa tried to solve the problem of still allowing manual downgrades in the Gemfile while `only_update_to_newer_versions` is true. But introduced a regression that prevented the `additional_base_requirements_for_resolve` method to work as intended:

This is the relevant change from that commit that tries to avoid adding the `>=` requirement if the  requirement in the Gemfile is different than the requirement in the lockfile (as far as I understand it):

```ruby
next requirements if @locked_deps[name] != dependencies_by_name[name]
```

I identified two problems
 1. `dependencies_by_name[name]` returns an array of `Bundler::Dependency`, where as 
`@locked_deps[name]` just returns a single `Bundler::Dependency`. Comparing the two will always be false.
 1. `@locked_deps` is always empty in case of `bundle update`. See: https://github.com/bundler/bundler/blob/3d9e6167a7df9ca89a030dfe95c7cdff293e74a9/lib/bundler/definition.rb#L95

### What is your fix for the problem, implemented in this PR?

My fixes:

 1. Make sure `dependencies_by_name` is a hash with `Bundler::Dependency` as values
 1. Fetch the `@locked_gems.dependencies` again instead of using `@locked_deps`
 1. The existing test worked for me with and without the `only_update_to_newer_versions` set to true, I replaced it with a reproduction of the edge case I was investigating (this is as minimal as I could make it)
 1. I've added a test for the manual downgrading case.

### Why did you choose this fix out of the possible options?

This is the only way I could make these cases work. It's possible there are other edge cases I don't understand.

### CODE REVIEW

File: lib/bundler/definition.rb [hunk 1]
```diff
@@ -974,10 +974,11 @@ def compute_requires
 
     def additional_base_requirements_for_resolve
       return [] unless @locked_gems && Bundler.feature_flag.only_update_to_newer_versions?
-      dependencies_by_name = dependencies.group_by(&:name)
+      dependencies_by_name = dependencies.each_with_object({}) {|dep, memo| memo[dep.name] = dep }
+      all_locked_deps = @locked_gems.dependencies
```
Comment (greysteil) [inline lib/bundler/definition.rb]: I'd rather lose this assignment and use the `@locked_gems.dependencies` directly in the loop

File: lib/bundler/definition.rb [hunk 2]
```diff
@@ -974,10 +974,11 @@ def compute_requires
 
     def additional_base_requirements_for_resolve
       return [] unless @locked_gems && Bundler.feature_flag.only_update_to_newer_versions?
-      dependencies_by_name = dependencies.group_by(&:name)
+      dependencies_by_name = dependencies.each_with_object({}) {|dep, memo| memo[dep.name] = dep }
```
Comment (greysteil) [inline lib/bundler/definition.rb]: We don't have `each_with_object` in Ruby 1.8.7, so this will need to use `reduce`

**Comment by theflow** [comment]
I've incorporated your comments

**Comment by greysteil** [comment]
@bundlerbot r+

**Comment by greysteil** [comment]
Thanks for this @theflow! 🎉
